### PR TITLE
Allow the tls module to work with Duplex streams as well

### PR DESCRIPTION
**Pull request by pmarkee**
IoT.js-DCO-1.0-Signed-off-by: Peter Marki marpeter@inf.u-szeged.hu

### CODE REVIEW

File: src/js/tls.js [hunk 1]
```diff
@@ -48,15 +48,20 @@ function TLSSocket(socket, options) {
...
   native.TlsInit(this, options, secureContext);
   this._socketState = socket._socketState;
+
+  if (socket._writableState.ready && !options.isServer) {
+    this.native_connect(options.servername || options.host || 'localhost');
+    this.native_read(null);
+  }
 }
-util.inherits(TLSSocket, EventEmitter);
+util.inherits(TLSSocket, stream.Duplex);
```
Comment (zherczeg) [inline src/js/tls.js]: I would investigate a bit more what is the best option here (mem/complexity).

File: src/js/tls.js [hunk 2]
```diff
@@ -48,15 +48,20 @@ function TLSSocket(socket, options) {
...
 
-TLSSocket.prototype._read = native.read;
-TLSSocket.prototype._write = native.write;
-TLSSocket.prototype._connect = native.connect;
+TLSSocket.prototype.native_read = native.read;
```
Comment (zherczeg) [inline src/js/tls.js]: `native_read` -> `_native_read` Rename the other three as well.

**Comment by zherczeg** [comment]
I have been thinking and there is no good way to avoid Duplex inheritance. So this is the best solution we have now.

File: src/js/tls.js [hunk 3]
```diff
@@ -15,7 +15,7 @@
 
 var net = require('net');
 var util = require('util');
-var EventEmitter = require('events').EventEmitter;
+var stream = require('stream');
```
Comment (zherczeg) [inline src/js/tls.js]: var Duplex = require('stream').Duplex;

**Comment by yichoi** [comment]
@pmarkee please update along with @zherczeg's comment

File: src/js/tls.js [hunk 4]
```diff
@@ -48,51 +50,40 @@ function TLSSocket(socket, options) {
...
-    if (!Buffer.isBuffer(data)) {
-      data = new Buffer(data);
-    }
-    data = this._write(data, true);
-  } else {
-    data = this._write(null, true);
-  }
-
-  this._socket.end(data, callback);
+  Writable.prototype.end.call(this, data, callback);
```
Comment (zherczeg) [inline src/js/tls.js]: I think Duplex.prototype.end would work here as well, and we could remove the Writable requirement.

File: src/js/tls.js [hunk 5]
```diff
@@ -48,51 +48,40 @@ function TLSSocket(socket, options) {
...
     this.on('secureConnect', callback);
   }
-
-  this._socket.connect(options);
-};
-
-TLSSocket.prototype.write = function(data, callback) {
-  if (!Buffer.isBuffer(data)) {
-    data = new Buffer(data);
+  if (this._socket instanceof net.Socket) {
```
Comment (zherczeg) [inline src/js/tls.js]: Do we need to check this? Connect should not be called unless `_socket` has a connect (tls.connect calls this).

File: src/js/tls.js [hunk 6]
```diff
@@ -48,51 +48,40 @@ function TLSSocket(socket, options) {
 
   native.TlsInit(this, options, secureContext);
   this._socketState = socket._socketState;
+
+  var self = this;
+  process.nextTick(function() {
+    if (socket._writableState.ready && !options.isServer) {
```
Comment (zherczeg) [inline src/js/tls.js]: I would check this first, and call `process.nextTick` if it is true.

File: src/js/tls.js [hunk 7]
```diff
@@ -212,10 +197,9 @@ function createServer(options, secureConnectionListener) {
 
 function connect(arg0, arg1, arg2, callback) {
   var options;
-  var tlsSocket;
   if (typeof arg0 == 'object') {
     options = Object.create(arg0, {
-      isServer: { value: false, enumerable: true },
+      isServer: { value: arg0.isServer || false, enumerable: true },
```
Comment (zherczeg) [inline src/js/tls.js]: `|| false` is pointless. Just let it be `arg0.isServer`.

File: src/js/tls.js [hunk 8]
```diff
@@ -257,8 +241,14 @@ function connect(arg0, arg1, arg2, callback) {
...
     }
   }
-  tlsSocket = new TLSSocket(new net.Socket(), options);
-  tlsSocket.connect(options, callback);
+
+  var tlsSocket = new TLSSocket(options.socket || new net.Socket(), options);
+  if (tlsSocket._socket instanceof net.Socket) {
+    tlsSocket.connect(options, callback);
+  }
+  if (util.isFunction(callback)) {
```
Comment (zherczeg) [inline src/js/tls.js]: Isn't this should be an `else if`?

File: src/js/tls.js (hunk 7 above)
Comment (zherczeg) [reply src/js/tls.js]: I just realiezd I don't understand this. How can a server connects???

Comment (pmarkee) [reply src/js/tls.js]: I left that in by accident, it doesn't have to be there

File: src/js/tls.js (hunk 8 above)
Comment (pmarkee) [reply src/js/tls.js]: I just realized you're right, yes, it has to be else if. The way I did it caused https tests to fail or timeout.

**Comment by yichoi** [comment]
LGTM

**Pull request merged**